dart/common/Addon.h

```diff
@@ -0,0 +1,305 @@
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class Properties : public Extensible<Properties> { };
+
+  /// Use the PropertiesMixer class to easily create a Properties extension
+  /// from an existing class or struct.
+  template <class Mixin>
+  using PropertiesMixer = ExtensibleMixer<Properties, Mixin>;
+
+  /// Virtual destructor
+  virtual ~Addon() = default;
+
+  /// Clone this Addon into a new manager
+  virtual std::unique_ptr<Addon> cloneAddon(AddonManager* newManager) const = 0;
+
+  /// Set the State of this Addon. By default, this does nothing.
+  virtual void setAddonState(const std::unique_ptr<State>& otherState);
+
+  /// Get the State of this Addon. By default, this returns a nullptr which
+  /// implies that the Addon is stateless.
+  virtual const State* getAddonState() const;
+
+  /// Set the Properties of this Addon. By default, this does nothing.
+  virtual void setAddonProperties(
+      const std::unique_ptr<Properties>& someProperties);
+
+  /// Get the Properties of this Addon. By default, this returns a nullptr
+  /// which implies that the Addon has no properties.
+  virtual const Properties* getAddonProperties() const;
+
+  /// This function will be called if the user is attempting to delete the Addon
+  /// but not immediately replacing it with another Addon of the same type. The
+  /// incoming argument will point to the AddonManager that had been holding
+  /// this Addon.
+  ///
+  /// If your Addon is mandatory for the AddonManager type that is passed in
+  /// here, then you should perform error handling in this function, and you
+  /// should return false to indicate that the operation is not permitted. If
+  /// you return false, then the Addon will NOT be removed from its Manager.
+  ///
+  /// By default, this simply returns true.
+  virtual bool isOptional(AddonManager* oldManager);
+
+protected:
+
+  /// Constructor
+  ///
+  // We require the AddonManager argument in this constructor to make it clear
+  // to extensions that they must have an AddonManager argument in their
+  // constructors.
+  Addon(AddonManager* manager);
+
+  /// This function should be overriden if your Addon needs to do any special
+  /// handling when its AddonManager gets changed.
+  virtual void changeManager(AddonManager* newManager);
+};
+
+//==============================================================================
+/// AddonWithProtectedState generates implementations of the State managing
+/// functions for an Addon class.
+template <class Base, typename StateData,
+          class ManagerType = AddonManager,
+          void (*updateState)(Base*) = &detail::NoOp<Base*> >
```

Comment:
[According to stackoverflow](http://stackoverflow.com/a/15090036), using a nullptr-valued function pointer is undefined behavior, so if we accept a nullptr template argument, then we would always have to check whether the value is a nullptr before using it. I'd much rather have a no-op function which (1) doesn't require special checking, and which (2) the compiler will optimize away entirely instead of having a branching if-statement (although since it's a ``constexpr``, the compiler might even be able to compile away any branching, so this could be a moot point).